src/riot-rs-threads/src/arch/cortex_m.rs

```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
 /// This function is called in PendSV.
 // TODO: make arch independent, or move to arch
 #[no_mangle]
-unsafe fn sched(old_sp: usize) -> usize {
-    let cs = CriticalSection::new();
-    let next_pid;
-
+unsafe fn sched(old_sp: usize) -> u128 {
     loop {
-        {
-            if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
-                next_pid = pid;
-                break;
+        if let Some(res) = critical_section::with(|cs| {
+            let threads = &mut *THREADS.as_ptr(cs);
+            if let Some(next_pid) = threads.runqueue.get_next() {
+                let current_high_regs = if let Some(current_pid) = threads.current_pid() {
+                    if next_pid == current_pid {
+                        return Some(0);
+                    }
+                    //println!("current: {} next: {}", current_pid, next_pid);
+                    threads.threads[current_pid as usize].sp = old_sp;
+                    threads.current_thread = Some(next_pid);
+                    threads.threads[current_pid as usize].data.as_ptr()
+                } else {
+                    core::ptr::null()
```

Comment:
I know that this is old code, but could we add a check in the asm code if r2 is `0` and skip the store commands in that case?

Reply:
that's already happening, if there's not "next_sp" (r0 == 0), the save/restore is skipped.

Reply:
`next_sp` is 0 if the next thread equals the current thread.
I was talking about the case when `r2` is `0` because there was no current thread before.

But I just noticed that right now we never have the case that `current_thread` is `None` for cortex-m, because we set it already in `start_threading`, before the first interrupt.
But as part of #182 I am proposing to change this.

Reply:
fixed in e3966d2864a57a49628b9742e31ac3a9ad8f218c.